multicorn_das/__init__.py

```diff
@@ -1,182 +1,367 @@
-            return
+    # ---------- gRPC call helpers ----------
 
-        log_to_postgres(f'Re-registering DAS with type: {self.fdw_options["das_type"]}', DEBUG)
+    @staticmethod
+    def _health_check(stub: HealthCheckServiceStub):
+        """
+        Try a gRPC health check to confirm the server is alive.
+        If the remote doesn't implement health-check, this may fail or raise UNIMPLEMENTED.
+        """
+        log_to_postgres('Performing health check...', DEBUG)
+        stub.Check(HealthCheckRequest())
+        log_to_postgres('Health check succeeded.', DEBUG)
+
+    @staticmethod
+    def _grpc_call_internal(das_name, das_type, das_url,
+                            stub_caller, request, *,
+                            attempts=30, health_stub=None,
+                            recreate_channel_callback=None,
+                            reregister_callback=None):
+        """
+        Unified wrapper for unary (non-streaming) gRPC calls.
+
+        Retry logic:
+          - If server is UNAVAILABLE, do a health check, re-create channel, and retry
+            up to `attempts` times. If fail, raise corresponding exception.
+          - If we get NOT_FOUND, call reregister_callback() ONCE, then retry.
+          - If UNAUTHENTICATION or PERMISSION_DENIED, raise corresponding error.
+          - Otherwise, raise a generic internal error.
+
+        Parameters:
+          stub_caller               A function that calls the correct method on the stub.
+          request                   The protobuf request object.
+          attempts                  Max attempts for UNAVAILABLE errors.
+          health_stub               If provided, do a health check with this stub on UNAVAILABLE.
+          recreate_channel_callback Called to re-create stubs if UNAVAILABLE is encountered.
+          reregister_callback       Called if NOT_FOUND is encountered (only once).
+
+        Returns:
+          The gRPC response from the stub method.
+        """
+        allow_reregister = True
 
-        registration_service = RegistrationServiceStub(self.channel)
+        for attempt in range(1, attempts + 1):
+            try:
+                return stub_caller(request)
+
+            except grpc.RpcError as e:
+                code = e.code()
+                log_to_postgres(f'gRPC error (attempt {attempt}/{attempts}): {code} - {e}', WARNING)
+
+                if code == grpc.StatusCode.UNAVAILABLE:
+                    # Out of attempts?
+                    if attempt == attempts:
+                        DASFdw._raise_unavailable(das_name, das_type, das_url, cause=e)
+                    # Sleep, recreate channel, health-check
+                    sleep(0.5)
+                    if recreate_channel_callback:
+                        recreate_channel_callback()
+                    if health_stub:
+                        try:
+                            DASFdw._health_check(health_stub)
```

Comment:
I don't see what difference it makes if health check succeeds or fails? If it fails we keep looping (`continue`), if it succeeds, we hit `continue` line 190. Right?

Reply:
Or could we just not call the health check endpoint and it would behave the same.